pkg/destroy/alibabacloud/alibabacloud.go

```diff
@@ -0,0 +1,1343 @@
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// New returns an Alibaba Cloud destroyer from ClusterMetadata.
+func New(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (providers.Destroyer, error) {
+	region := metadata.ClusterPlatformMetadata.AlibabaCloud.Region
+	client, err := icalibabacloud.NewClient(region)
+	if err != nil {
+		return nil, err
+	}
+
+	auth := credentials.NewAccessKeyCredential(client.AccessKeyID, client.AccessKeySecret)
+
+	return &ClusterUninstaller{
+		Logger:          logger,
+		Auth:            auth,
+		AccessKeyID:     client.AccessKeyID,
+		AccessKeySecret: client.AccessKeySecret,
+		Region:          region,
+		InfraID:         metadata.InfraID,
+		ClusterDomain:   metadata.AlibabaCloud.ClusterDomain,
+		Tags: []map[string]string{
+			{
+				fmt.Sprintf("kubernetes.io/cluster/%s", metadata.InfraID): "owned",
+			},
+			{
+				"ack.aliyun.com": metadata.InfraID,
+			},
+		},
+	}, nil
+}
+
+// Run is the entrypoint to start the uninstall process.
+func (o *ClusterUninstaller) Run() (*types.ClusterQuota, error) {
+	var err error
+
+	err = o.configureClients()
+	if err != nil {
+		return nil, err
+	}
+
+	err = o.findResources()
+	if err != nil {
+		return nil, err
+	}
+
+	err = o.destroyCluster()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to destroy cluster")
+	}
+
+	return nil, nil
+}
+
+func (o *ClusterUninstaller) destroyCluster() error {
+	stagedFuncs := [][]struct {
+		name    string
+		execute func() error
+	}{
+		{
+			{name: "DNS records", execute: o.deleteDNSRecords},
+			{name: "OSS buckets", execute: o.deleteBucket},
+			{name: "RAM roles", execute: o.deleteRAMRoles},
+			{name: "ECS instances", execute: o.deleteEcsInstances},
+		},
+		{
+			{name: "private zones", execute: o.deletePrivateZones},
+			{name: "ECS security groups", execute: o.deleteSecurityGroups},
+			{name: "Nat gateways", execute: o.deleteNatGateways},
+			{name: "SLBs", execute: o.deleteSlbs},
+		},
+		{
+			{name: "EIPs", execute: o.deleteEips},
+		},
+		{
+			{name: "VSwitchs", execute: o.deleteVSwitchs},
+		},
+		{
+			{name: "VPCs", execute: o.deleteVpcs},
+		},
+		{
+			{name: "resource groups", execute: o.deleteResourceGroup},
+		},
+	}
+
+	for _, stage := range stagedFuncs {
+		var wg sync.WaitGroup
+		errCh := make(chan error)
+		wgDone := make(chan bool)
+
+		for _, f := range stage {
+			wg.Add(1)
+			go o.executeStageFunction(f, errCh, &wg)
+		}
+
+		go func() {
+			wg.Wait()
+			close(wgDone)
+		}()
+
+		select {
+		case <-wgDone:
+			// On to the next stage
+			continue
+		case err := <-errCh:
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (o *ClusterUninstaller) executeStageFunction(f struct {
+	name    string
+	execute func() error
+}, errCh chan error, wg *sync.WaitGroup) error {
+	defer wg.Done()
+
+	err := wait.PollImmediateInfinite(
+		time.Second*10,
+		func() (bool, error) {
+			ferr := f.execute()
```

Comment:
I do not love this solution and I am not proposing it but this addition was able to skip the resources that are missing:
```
			if ferr != nil {
				if strings.Contains(ferr.Error(), "not exist") {
					return true, nil
				}
				o.Logger.Debugf("%s: %v", f.name, ferr)
				return false, nil
			}
```

I recommend using something similar to this to skip waiting if an object returns `NoExist`.
@bd233 @patrickdillon WDYT?

Reply:
Is the `deleteDNSRecords` function incorrectly returning an error when it can't find the domain?

Reply:
The `deleteDNSRecords` is returning an error that the DNS record does not exist. I think this behavior is expected from the Alibaba API. We need to gracefully handle `NoExist` errors so that deletion can complete successfully.

@bd233 Can you confirm this behavior?

Reply:
Another option is to check to see if it exists before trying to delete. That is what I thought this code was doing: https://github.com/openshift/installer/pull/5348/files#diff-5d31e7ab73dc252cc09331a0790fcc8b6cd3add977402effbe19a1b1b24fba39R1249-R1255

Reply:
I'm sorry to reply so late

Reply:
@kwoodson @patrickdillon  Sorry to reply so late.
I think this is an abnormal scenario. I read the `Start to search DNS records` log, but did not see `Start to delete DNS records`, so the error should come from `records, err: = o.listrecord (basedomain)`, But this should not be, if the domain name does not exist, it should be returned here:

> 	 if len(domains) == 0 {
> 		 return
> 	}

I am trying to reproduce and fix this problem as soon as possible.

>  I recommend checking the ferr message for objects that do not exist and return true to continue the deletion.
>  @bd233 @patrickdillon WDYT?

I don't think 'ferr' should be handled. This error should be resolved in 'deleteDnsRecords'. This solution may cause some resources not to be deleted

@kwoodson Are there any manual deletion operations in the process of executing the installer？

Reply:
@bd233 Generally speaking there shouldn't be any but that doesn't stop customers from updating or modifying the cluster objects. The reason this DNS item was modified is that we set up valid DNS and removed the original before removing the cluster. That caused the missing DNS record. I think that it is not unreasonable to have components get modified or deleted outside of the cluster. I think if objects return `NoExist` they should be considered to be deleted and the destroy code can continue.

Reply:
@kwoodson @patrickdillon 
I think I understand this problem. Querying domainName is a fuzzy query. If baseDomain is `openshift.com`, then` bd.openshift.com` may appear in the result, and `len (domains) == 0 ` is false, so...

I'll fix it as soon as possible.

Reply:
I think `ListPrivateZones` and `ListDNSDomain` in `pkg/asset/installconfig/alibabacloud/client.go` also has this problem. I will create a PR to fix it.